Sam guide component docs (#1017)
knocklabs/docs: content/in-app-ui/react/modal.mdx

<Callout
  emoji="⚠️"
  text={
    <>
      <span className="font-bold">Note:</span> You must be on @knocklabs/react
      version 0.7.0 or higher to use guides-related features.
    </>
  }
/>

### Installing dependencies

```bash title="Installing dependencies"
npm install @knocklabs/react
```

## Basic usage

<Steps titleSize="h3">
  <Step title="Setup the KnockGuideProvider">
    
    First, wrap your application with the `KnockProvider` and `KnockGuideProvider`. The `KnockGuideProvider` requires a `channelId`. You can find your guide channel ID on the integrations page in the dashboard under "Channels".

    ```tsx title="Setup the KnockGuideProvider within your product."
    import { KnockProvider, KnockGuideProvider } from "@knocklabs/react";
    import { useCurrentUser } from "@/lib/hooks";

    const MyApplication = () => {
      // Get your authenticated current user
      const currentUser = useCurrentUser();

      return (
        <KnockProvider
          apiKey={process.env.NEXT_PUBLIC_KNOCK_API_KEY}
          userId={currentUser.id}
        >
          <KnockGuideProvider
            channelId={process.env.NEXT_PUBLIC_KNOCK_GUIDE_CHANNEL_ID}
          >

>             <Header />
>             <Sidebar />
>             {/* Rest of your app */}

Reviewer: ```suggestion
            <MyLayout>
              {/* Rest of your app */}
            </MyLayout>
```
Just a thought that I had, whether it'd be good to reference the component we are using as an example below.

          </KnockGuideProvider>
        </KnockProvider>
      );
    };
    ```

  </Step>
  <Step title="Add the Modal component">
    

>     Import Knock's pre-built `Modal` component and place it under the `KnockGuideProvider` near the top of your application. The `<Modal/>` component is most effective when placed in the root layout of your application, this ensures that your modal is available to render on every page of your application.

Reviewer: ```suggestion
    Import Knock's pre-built `Modal` component and place it inside the `KnockGuideProvider` near the top of your application's component tree. The `<Modal/>` component is most effective when placed in the root layout of your application, this ensures that your modal is available to render on every page of your application.
```

    ```tsx title="Add the modal near the top of your application."
    import { Modal } from "@knocklabs/react";

>     const DashboardLayout = ({children}) => {

Reviewer: ```suggestion
    const MyLayout = ({children}) => {
```

    You can learn more about guide targeting and activation rules in the [creating guides](/in-app-ui/guides/create-guides) page.

  </Step>
</Steps>

## Working with modal variants

The pre-built modal message type supports three variants for different use cases:

- **Default.** A dismissible modal with no action buttons.
- **Single action.** A dismissible modal with a single action button.
- **Multi-action.** A dismissible modal with two action buttons.

## Handling user engagement

The modal component handles user engagement tracking automatically. Here's an overview of what user behavior maps to which engagement statuses tracked in Knock.

- **Seen.** The modal has been rendered to (seen by) the user.
- **Interacted.** The user has interacted with (clicked) the modal. Dismissing the modal does not count as an interaction.
- **Archived.** The modal has been archived (dismissed) by the user.

## Styling your modal

### CSS customization

You can style your modal using standard CSS. Here's a basic example:

```css title="Modal styles"

> .modal-overlay {

Reviewer: The class names here are not quite right. [Here](https://github.com/knocklabs/javascript/blob/18be7a567363c1056b6f420aa45c6d83643f49ed/packages/react/src/modules/guide/components/Modal/styles.css) is what we have in the latest SDK we can reference. 

One other thing to touch on in this "Styling" section is that we use [css variables](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_cascading_variables/Using_CSS_custom_properties) (e.g. `--knock-guide-accent`) in our css definitions (linked above). This was intentional as we wanted to make it easy to "theme" the prebuilt components using your own color tokens if you are otherwise happy with how the components are styled. More specifically, the ones prefixed with `--knock-guide-` are meant to be overridden for the theming purposes. 

So there are different layers of abstractions we provide that you can use to approach styling, from the simplest (but limited) to most involved (but flexible):
* Override the `--knock-guide-` css vars to "theme" the components with your own colors
* Override some or all of the css definitions to completely customize the styles
* Use the individual [sub components](https://github.com/knocklabs/javascript/blob/18be7a567363c1056b6f420aa45c6d83643f49ed/packages/react/src/modules/guide/components/Modal/Modal.tsx#L339-L351) to compose your own component (and use inline style props, if needed)
  * This is technically available but i think at this point most people would reach for the BYOC option
* "Bring your own component (BYOC)" and style freely, and leverage the useGuide hook to power it.

  padding: 24px;
  max-width: 500px;
  width: 90%;
  max-height: 80vh;
  overflow-y: auto;
  box-shadow: 0 10px 25px rgba(0, 0, 0, 0.2);
}

.modal-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 16px;
}

.modal-body {
  margin-bottom: 24px;
}

.modal-footer {
  display: flex;
  justify-content: flex-end;
  gap: 8px;
}

.close-button {
  background: none;
  border: none;
  font-size: 24px;
  cursor: pointer;
  color: #666;
}

.primary-button {
  background: #007bff;
  color: white;
  border: none;
  padding: 8px 16px;
  border-radius: 4px;
  cursor: pointer;
}

.secondary-button {
  background: transparent;
  color: #007bff;
  border: 1px solid #007bff;
  padding: 8px 16px;
  border-radius: 4px;
  cursor: pointer;
}
```

## Build your own modal component

You can build your own modal component using the `useGuide` hook.

```tsx title="Building your own modal component"
import { useEffect } from "react";
import { useGuide } from "@knocklabs/react";

const ByoModal = () => {
  const { step } = useGuide({ type: "modal" });

  useEffect(() => {
    if (step) step.markAsSeen();
  }, [step]);

  if (!step) return null;

  return (

>     <div onClick={() => step.markAsInteracted()}>

Reviewer: I think it'd be better to show an example with a button and `step.markAsInteracted()` as an onClick handler for the button rather than at the root container, because this markup structure can lead to firing markAsInteracted when a dismiss button is clicked inside, for example. You can prevent that with `event.preventDefault()` and/or `event.stopPropagation()` but that's more things to explain.

```
{step.content.primary_button && (
  <button onClick={() => step.markAsInteracted()}>...</button>
)}
```
